ICSharpCode.Decompiler/IL/Transforms/EarlyExpressionTransforms.cs

```diff
@@ -111,5 +111,49 @@ internal static ILInstruction HandleCall(Call inst, ILTransformContext context)
 			}
 			return null;
 		}
+
+		protected internal override void VisitBinaryNumericInstruction(BinaryNumericInstruction inst)
+		{
+			if (inst.Operator != BinaryNumericOperator.BitAnd)
+				return;
+
+			if (PotentialSideEffects(inst.Left) || PotentialSideEffects(inst.Right))
+				return;
+
+			if (!IsBoolean(inst.Left) || !IsBoolean(inst.Right))
+				return;
+			
+			context.Step("Replace bit.and with logic.and", inst);
+			var expr = IfInstruction.LogicAnd(inst.Left, inst.Right);
+			inst.ReplaceWith(expr);
+			expr.AcceptVisitor(this);
+		}
+
+		private bool IsBoolean(ILInstruction inst) => inst is Comp || inst.InferType().IsKnownType(KnownTypeCode.Boolean);
+
+		private bool PotentialSideEffects(ILInstruction inst)
```

Comment:
This function already exists: `!SemanticHelper.IsPure(inst.Flags)`
(your version misses that `BinaryNumericInstruction` can have side effects, e.g. `OverflowException` or `DivideByZeroException`)